securedrop/tests/functional/functional_test.py

```diff
@@ -33,20 +34,69 @@ class FunctionalTest():
 
     def _unused_port(self):
         s = socket.socket()
-        s.bind(("localhost", 0))
+        s.bind(('127.0.0.1', 0))
         port = s.getsockname()[1]
         s.close()
         return port
 
-    def _create_webdriver(self):
-        log_file = open(join(LOG_DIR, 'firefox.log'), 'a')
-        log_file.write(
-            '\n\n[%s] Running Functional Tests\n' % str(
-                datetime.now()))
-        log_file.flush()
+
+    def _create_tor_browser_webdriver(self, abs_log_file_path):
+        # Creating the TorBrowser driver changes the working directory,
+        # so we need to save and then restore the current one.
+        old_dir = os.getcwd()
+
+        # Read the TorBrowser location from Ansible
+        with open('/opt/.tbb_path_file') as f:
+           path_to_tbb = f.readline().strip()
+        path_to_tbb = path_to_tbb + os.path.sep + "tor-browser_en-US"
+        path_to_tbb = os.path.abspath(path_to_tbb)
+
+        # Due to Travis having opinions, we need to chown the tor dir
+        # to be owned by root when we're running CI.
+        if '/home/travis/' in path_to_tbb:
+            for root, dirs, files in os.walk(path_to_tbb):
+                for d in dirs:
+                    os.chown(os.path.join(root, d), 0, 0)
+                for f in files:
+                    os.chown(os.path.join(root, f), 0, 0)
+
+        # Don't use Tor when reading from localhost,
+        # and turn off private browsing.
+        # We need this to make functional tests work,
```

Comment:
Be more explicit here. We need to turn off private browsing because we won't be able to access the browser's cookies in private browsing mode. Since we use session cookies in SD anyway (in private browsing mode all cookies are set as session cookies), this should not affect session lifetime.